#### Import libraries

In [1]:
import pandas as pd

## Reading in a data file

Read in the data file: **df**

In [2]:
data_file = './datasets/1981-2010 NOAA Austin Climate Normals.csv' 
df = pd.read_csv(data_file)
df.head()

13904  20110101  0053  12  OVC045     10.00  .1  .2  .3 ...   .18  .19  \
0  13904  20110101   153  12  OVC049     10.00             ...              
1  13904  20110101   253  12  OVC060     10.00             ...   030        
2  13904  20110101   353  12  OVC065     10.00             ...              
3  13904  20110101   453  12  BKN070     10.00             ...              
4  13904  20110101   553  12  BKN065     10.00             ...   015        

   29.95  .20  AA  .21  .22  .23 29.95.1  .24  
0  30.01       AA                  30.02       
1  30.01       AA                  30.02       
2  30.03       AA                  30.04       
3  30.04       AA                  30.04       
4  30.06       AA                  30.06       

[5 rows x 44 columns]

Read in the data file with **header=None**: **df_headers**

In [3]:
df_headers = pd.read_csv(data_file, header=None)
df_headers.head()

0         1    2   3       4  5      6  7  8  9  ...   34 35     36 37  \
0  13904  20110101   53  12  OVC045     10.00          ...          29.95      
1  13904  20110101  153  12  OVC049     10.00          ...          30.01      
2  13904  20110101  253  12  OVC060     10.00          ...  030     30.01      
3  13904  20110101  353  12  OVC065     10.00          ...          30.03      
4  13904  20110101  453  12  BKN070     10.00          ...          30.04      

   38 39 40 41     42 43  
0  AA           29.95     
1  AA           30.02     
2  AA           30.02     
3  AA           30.04     
4  AA           30.04     

[5 rows x 44 columns]

## Re-assigning column names

Split on the comma to create a list **column_labels_list**. After that, assign the new column labels to the DataFrame **df.columns**

In [4]:
column_labels = 'Wban,date,Time,StationType,sky_condition,sky_conditionFlag,visibility,visibilityFlag,wx_and_obst_to_vision,wx_and_obst_to_visionFlag,dry_bulb_faren,dry_bulb_farenFlag,dry_bulb_cel,dry_bulb_celFlag,wet_bulb_faren,wet_bulb_farenFlag,wet_bulb_cel,wet_bulb_celFlag,dew_point_faren,dew_point_farenFlag,dew_point_cel,dew_point_celFlag,relative_humidity,relative_humidityFlag,wind_speed,wind_speedFlag,wind_direction,wind_directionFlag,value_for_wind_character,value_for_wind_characterFlag,station_pressure,station_pressureFlag,pressure_tendency,pressure_tendencyFlag,presschange,presschangeFlag,sea_level_pressure,sea_level_pressureFlag,record_type,hourly_precip,hourly_precipFlag,altimeter,altimeterFlag,junk'
column_labels_list = column_labels.split(',')
df.columns = column_labels_list

Remove the appropriate columns: **df_dropped**

In [5]:
list_to_drop = ['sky_conditionFlag', 'visibilityFlag', 'wx_and_obst_to_vision', 'wx_and_obst_to_visionFlag', 'dry_bulb_farenFlag', 'dry_bulb_celFlag', 'wet_bulb_farenFlag', 'wet_bulb_celFlag', 'dew_point_farenFlag', 'dew_point_celFlag', 'relative_humidityFlag', 'wind_speedFlag', 'wind_directionFlag', 'value_for_wind_character', 'value_for_wind_characterFlag', 'station_pressureFlag', 'pressure_tendencyFlag', 'pressure_tendency', 'presschange', 'presschangeFlag', 'sea_level_pressureFlag', 'hourly_precip', 'hourly_precipFlag', 'altimeter', 'record_type', 'altimeterFlag', 'junk']
df_dropped = df.drop(list_to_drop, axis='columns')
df_dropped.head()

Wban      date  Time  StationType sky_condition visibility dry_bulb_faren  \
0  13904  20110101   153           12        OVC049      10.00             51   
1  13904  20110101   253           12        OVC060      10.00             51   
2  13904  20110101   353           12        OVC065      10.00             50   
3  13904  20110101   453           12        BKN070      10.00             50   
4  13904  20110101   553           12        BKN065      10.00             49   

  dry_bulb_cel wet_bulb_faren wet_bulb_cel dew_point_faren dew_point_cel  \
0         10.6             37          3.0              14         -10.0   
1         10.6             37          2.9              13         -10.6   
2         10.0             38          3.1              17          -8.3   
3         10.0             37          2.8              15          -9.4   
4          9.4             37          2.8              17          -8.3   

  relative_humidity wind_speed wind_direction station_pressure  \
0                23         10            340            29.49   
1                22         15            010            29.49   
2                27          7            350            29.51   
3                25         11            020            29.51   
4                28          6            010            29.53   

  sea_level_pressure  
0              30.01  
1              30.01  
2              30.03  
3              30.04  
4              30.06

## Cleaning and tidying datetime data

Convert the date column to string: **df_dropped['date']**

In [6]:
df_dropped['date'] = df_dropped['date'].astype(str)

Pad leading zeros to the Time column: **df_dropped['Time']**

In [7]:
df_dropped['Time'] = df_dropped['Time'].apply(lambda x:'{:0>4}'.format(x))

Concatenate the new date and Time columns: **date_string**

In [8]:
date_string = df_dropped['date'] + df_dropped['Time']
date_string.head()

0    201101010153
1    201101010253
2    201101010353
3    201101010453
4    201101010553
dtype: object

Convert the date_string Series to datetime: **date_times**

In [9]:
date_times = pd.to_datetime(date_string, format='%Y%m%d%H%M')
date_times.head()

0   2011-01-01 01:53:00
1   2011-01-01 02:53:00
2   2011-01-01 03:53:00
3   2011-01-01 04:53:00
4   2011-01-01 05:53:00
dtype: datetime64[ns]

Set the index to be the new date_times container: **df_clean**

In [10]:
df_clean = df_dropped.set_index(date_times)
df_clean.head()

Wban      date  Time  StationType sky_condition  \
2011-01-01 01:53:00  13904  20110101  0153           12        OVC049   
2011-01-01 02:53:00  13904  20110101  0253           12        OVC060   
2011-01-01 03:53:00  13904  20110101  0353           12        OVC065   
2011-01-01 04:53:00  13904  20110101  0453           12        BKN070   
2011-01-01 05:53:00  13904  20110101  0553           12        BKN065   

                    visibility dry_bulb_faren dry_bulb_cel wet_bulb_faren  \
2011-01-01 01:53:00      10.00             51         10.6             37   
2011-01-01 02:53:00      10.00             51         10.6             37   
2011-01-01 03:53:00      10.00             50         10.0             38   
2011-01-01 04:53:00      10.00             50         10.0             37   
2011-01-01 05:53:00      10.00             49          9.4             37   

                    wet_bulb_cel dew_point_faren dew_point_cel  \
2011-01-01 01:53:00          3.0              14         -10.0   
2011-01-01 02:53:00          2.9              13         -10.6   
2011-01-01 03:53:00          3.1              17          -8.3   
2011-01-01 04:53:00          2.8              15          -9.4   
2011-01-01 05:53:00          2.8              17          -8.3   

                    relative_humidity wind_speed wind_direction  \
2011-01-01 01:53:00                23         10            340   
2011-01-01 02:53:00                22         15            010   
2011-01-01 03:53:00                27          7            350   
2011-01-01 04:53:00                25         11            020   
2011-01-01 05:53:00                28          6            010   

                    station_pressure sea_level_pressure  
2011-01-01 01:53:00            29.49              30.01  
2011-01-01 02:53:00            29.49              30.01  
2011-01-01 03:53:00            29.51              30.03  
2011-01-01 04:53:00            29.51              30.04  
2011-01-01 05:53:00            29.53              30.06

## Cleaning the numeric columns

Print the **dry_bulb_faren** temperature between **8 AM and 9 AM on June 20, 2011** using **_partial datetime string selection_**.

In [11]:
print(df_clean.loc['2011-06-20 08:00:00':'2011-06-20 09:00:00', 'dry_bulb_faren'])

2011-06-20 08:27:00     M
2011-06-20 08:28:00     M
2011-06-20 08:29:00     M
2011-06-20 08:30:00     M
2011-06-20 08:31:00     M
2011-06-20 08:32:00     M
2011-06-20 08:33:00     M
2011-06-20 08:34:00     M
2011-06-20 08:35:00     M
2011-06-20 08:53:00    83
Name: dry_bulb_faren, dtype: object


Convert the **dry_bulb_faren** column to numeric values: **df_clean['dry_bulb_faren']**

Obs.: By specifying the keyword argument **errors='coerce'**, you can force strings like **'M'** to be interpreted as **NaN**

In [13]:
df_clean['dry_bulb_faren'] = pd.to_numeric(df_clean['dry_bulb_faren'], errors='coerce')

Print the transformed **dry_bulb_faren** temperature between **8 AM and 9 AM on June 20, 2011**

In [14]:
print(df_clean.loc['2011-06-20 08:00:00':'2011-06-20 09:00:00', 'dry_bulb_faren'])

2011-06-20 08:27:00     NaN
2011-06-20 08:28:00     NaN
2011-06-20 08:29:00     NaN
2011-06-20 08:30:00     NaN
2011-06-20 08:31:00     NaN
2011-06-20 08:32:00     NaN
2011-06-20 08:33:00     NaN
2011-06-20 08:34:00     NaN
2011-06-20 08:35:00     NaN
2011-06-20 08:53:00    83.0
Name: dry_bulb_faren, dtype: float64


Convert the **wind_speed** and **dew_point_faren** columns to numeric values

In [15]:
df_clean['wind_speed'] = pd.to_numeric(df_clean['wind_speed'], errors='coerce')
df_clean['dew_point_faren'] = pd.to_numeric(df_clean['dew_point_faren'], errors='coerce')

## Statistical EDA (Exploratory Data Analysis)

### 1. Median

Print the median of the dry_bulb_faren column

In [18]:
print(df_clean['dry_bulb_faren'].median())

72.0


Print the median of the dry_bulb_faren column for the time range **'2011-Apr':'2011-Jun'**

In [20]:
print(df_clean.loc['2011-Apr':'2011-Jun', 'dry_bulb_faren'].median())

78.0


Print the median of the dry_bulb_faren column for the month of **January**

In [21]:
print(df_clean.loc['2011-Jan', 'dry_bulb_faren'].median())

48.0


### 2. Variance

Read in the data file: **df_climate**

In [52]:
data_austin_weather_2010 = './datasets/2010 Austin weather.csv' 
df_climate = pd.read_csv(data_austin_weather_2010)
df_climate.head()

Temperature  DewPoint  Pressure            Date
0         46.2      37.5       1.0  20100101 00:00
1         44.6      37.1       1.0  20100101 01:00
2         44.1      36.9       1.0  20100101 02:00
3         43.8      36.9       1.0  20100101 03:00
4         43.5      36.8       1.0  20100101 04:00

Convert the **Date Column** of df_climate to **datetime*: **date_times**

In [58]:
date_times = pd.to_datetime(df_climate['Date'], format='%Y%m%d %H:%M')
df_climate = df_climate.drop('Date', axis='columns')
df_climate = df_climate.set_index(date_times)
df_climate.head()

Temperature  DewPoint  Pressure
Date                                                
2010-01-01 00:00:00         46.2      37.5       1.0
2010-01-01 01:00:00         44.6      37.1       1.0
2010-01-01 02:00:00         44.1      36.9       1.0
2010-01-01 03:00:00         43.8      36.9       1.0
2010-01-01 04:00:00         43.5      36.8       1.0

Downsample **df_clean** by **day** and aggregate by **mean**: **daily_mean_2011**

In [59]:
daily_mean_2011 = df_clean.resample('D').mean()
daily_mean_2011.head()

Wban  StationType  dry_bulb_faren  dew_point_faren  wind_speed
2011-01-01  13904           12       50.130435        20.739130   10.913043
2011-01-02  13904           12       39.416667        19.708333    4.166667
2011-01-03  13904           12       46.846154        35.500000    2.653846
2011-01-04  13904           12       53.367347        50.408163    2.510204
2011-01-05  13904           12       57.965517        40.068966    4.689655

Downsample **df_clean** by **day** and aggregate by **mean**: **daily_mean_2011**

In [60]:
daily_mean_2011 = df_clean.resample('D').mean()

Extract the **dry_bulb_faren** column from **daily_mean_2011** using **.values**: **daily_temp_2011**

In [61]:
daily_temp_2011 = daily_mean_2011['dry_bulb_faren'].values

Downsample **df_climate** by **day** and aggregate by **mean**: **daily_climate**

In [62]:
daily_climate = df_climate.resample('D').mean()

Extract the **Temperature** column from **daily_climate** using **.reset_index()**: **daily_temp_climate**

In [66]:
daily_temp_climate = daily_climate.reset_index()['Temperature']
daily_temp_climate.head()

0    49.337500
1    49.795833
2    49.900000
3    49.729167
4    49.841667
Name: Temperature, dtype: float64

Compute the difference between the two arrays and print **the mean difference**

In [67]:
difference = daily_temp_2011 - daily_temp_climate
print(difference.mean())

1.330083921569873


## Sunny or cloudy

**On average, how much hotter is it when the sun is shining?**

The column **'sky_condition'** provides information about whether the day was **sunny** ('CLR') or **overcast** ('OVC').

Get the maximum temperatures when sky_condition is **sunny** ('CLR')

In [68]:
is_sky_clear = df_clean['sky_condition']=='CLR'
sunny = df_clean[is_sky_clear]
sunny_daily_max = sunny.resample('D').max()
sunny_daily_max.head()

Wban      date  Time  StationType sky_condition  \
2011-01-01  13904.0  20110101  2353         12.0           CLR   
2011-01-02  13904.0  20110102  2253         12.0           CLR   
2011-01-03  13904.0  20110103  0453         12.0           CLR   
2011-01-04      NaN       NaN   NaN          NaN           NaN   
2011-01-05  13904.0  20110105  2353         12.0           CLR   

            dry_bulb_faren dry_bulb_cel wet_bulb_faren wet_bulb_cel  \
2011-01-01            59.0          8.3             45          7.2   
2011-01-02            35.0          1.7             32          0.1   
2011-01-03            32.0          0.0             29         -1.9   
2011-01-04             NaN          NaN            NaN          NaN   
2011-01-05            35.0          1.7             33          0.3   

            dew_point_faren dew_point_cel relative_humidity  wind_speed  \
2011-01-01             28.0          -6.1                53        16.0   
2011-01-02             28.0          -7.2                76         8.0   
2011-01-03             26.0          -4.4                85         0.0   
2011-01-04              NaN           NaN               NaN         NaN   
2011-01-05             29.0          -1.7                79         0.0   

           wind_direction station_pressure sea_level_pressure  
2011-01-01            360            29.78              30.33  
2011-01-02            360            29.82              30.38  
2011-01-03            000            29.71              30.27  
2011-01-04            NaN              NaN                NaN  
2011-01-05            000            29.54              30.08

Get the maximum temperatures when sky_condition is **overcast** ('OVR')

In [69]:
is_sky_overcast = df_clean['sky_condition'].str.contains('OVC')
overcast = df_clean[is_sky_overcast]
overcast_daily_max = overcast.resample('D').max()
overcast_daily_max.head()

Wban      date  Time  StationType  sky_condition  \
2011-01-01  13904.0  20110101  0353         12.0         OVC065   
2011-01-02      NaN       NaN   NaN          NaN            NaN   
2011-01-03  13904.0  20110103  2353         12.0  SCT042 OVC055   
2011-01-04  13904.0  20110104  2353         12.0  SCT010 OVC016   
2011-01-05  13904.0  20110105  0653         12.0  SCT006 OVC011   

            dry_bulb_faren dry_bulb_cel wet_bulb_faren wet_bulb_cel  \
2011-01-01            51.0         10.6             38          3.1   
2011-01-02             NaN          NaN            NaN          NaN   
2011-01-03            58.0          9.4             49          9.7   
2011-01-04            57.0          8.9             56          9.4   
2011-01-05            57.0         14.0             56         13.5   

            dew_point_faren dew_point_cel relative_humidity  wind_speed  \
2011-01-01             17.0          -8.3                27        15.0   
2011-01-02              NaN           NaN               NaN         NaN   
2011-01-03             45.0           7.0                79        10.0   
2011-01-04             56.0           8.9               100         8.0   
2011-01-05             56.0          13.3                96         3.0   

           wind_direction station_pressure sea_level_pressure  
2011-01-01            350            29.51              30.03  
2011-01-02            NaN              NaN                NaN  
2011-01-03            200            29.70                  M  
2011-01-04            VR             29.59                  M  
2011-01-05            250            29.48                  M

In [ ]:
Calculate the mean of sunny_daily_max and overcast_daily_max. Next

In [70]:
# Calculate the mean of sunny_daily_max
sunny_daily_max_mean = sunny_daily_max.mean()

# Calculate the mean of overcast_daily_max
overcast_daily_max_mean = overcast_daily_max.mean()

# Print the difference (sunny minus overcast)
difference = sunny_daily_max_mean - overcast_daily_max_mean
print(difference)

Wban               0.000000
StationType        0.000000
dry_bulb_faren     6.504304
dew_point_faren   -4.339286
wind_speed        -3.246062
dtype: float64


If you see in the result above, the average daily maximum dry bulb temperature was 6.5 degrees Fahrenheit higher on sunny days compared to overcast days. It means that in sunny days tha temperature was higher than overcast days in this year

-------------------------------------------------------------------------------------------------------------------